src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.XPlat/UpdateNuGetConfigPackageSourcesMappings.cs

```diff
@@ -85,5 +169,149 @@ public override bool Execute()
+            foreach (string nuspecFile in nuspecFiles)
+            {
+                try
+                {
+                    using Stream stream = File.OpenRead(nuspecFile);
+                    NupkgInfo info = GetNupkgInfo(stream);
+                    string id = info.Id.ToLower();
+                    string version = info.Version.ToLower();
+
+                    AddToHashtable(currentPackages, id, version);
+                    AddToHashtable(allSourcesPackages, SBRPCacheSourceName, id);
+                }
+                catch (Exception ex)
+                {
+                    throw new InvalidDataException(string.Format(CultureInfo.CurrentCulture, "Invalid nuspec file", nuspecFile), ex);
+                }
+            }
+        }
+
+        private void AddToHashtable(Hashtable hashtable, string key, string value)
+        {
+            if (hashtable.ContainsKey(key))
+            {
+                List<string> values = (List<string>)hashtable[key];
+                if (!values.Contains(value))
+                {
+                    values.Add(value);
+                }
+            }
+            else
+            {
+                hashtable.Add(key, new List<string> { value });
+            }
+        }
+
+        /// <summary>
+        /// Get nupkg info, id and version, from nupkg file.
+        /// </summary>
+        private NupkgInfo GetNupkgInfo(string path)
+        {
+            try
+            {
+                using Stream stream = File.OpenRead(path);
+                ZipArchive zipArchive = new(stream, ZipArchiveMode.Read);
+                foreach (ZipArchiveEntry entry in zipArchive.Entries)
+                {
+                    if (entry.Name.EndsWith(".nuspec"))
+                    {
+                        using Stream nuspecFileStream = entry.Open();
+                        return GetNupkgInfo(nuspecFileStream);
+                    }
+                }
+
+                throw new InvalidDataException(string.Format(CultureInfo.CurrentCulture, "Did not extract nuspec file from package: {0}", path));
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidDataException(string.Format(CultureInfo.CurrentCulture, "Invalid package", path), ex);
+            }
+        }
+
+        /// <summary>
+        /// Get nupkg info, id and version, from nuspec stream.
+        /// </summary>
+        private NupkgInfo GetNupkgInfo(Stream nuspecFileStream)
+        {
+            XDocument doc = XDocument.Load(nuspecFileStream, LoadOptions.PreserveWhitespace);
+            XElement metadataElement = doc.Descendants().First(c => c.Name.LocalName.ToString() == "metadata");
+            return new NupkgInfo(
+                    metadataElement.Descendants().First(c => c.Name.LocalName.ToString() == "id").Value,
+                    metadataElement.Descendants().First(c => c.Name.LocalName.ToString() == "version").Value);
+        }
+    }
+
+    public class NupkgInfo
```

Comment:
Consider making a private nested type as it should not be used outside the task.

Reply:
Fixed with https://github.com/dotnet/installer/pull/19114/commits/3f02bc95f3815f096425785f6fd3e2a351888a7e